src/sdk/pynni/nni/curvefitting_assessor/curvefitting_assessor.py

```diff
@@ -0,0 +1,78 @@
+#
+# THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING
+# BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+# DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+import logging
+import numpy as np
+from nni.assessor import Assessor, AssessResult
+from .modelfactory import CurveModel
+
+logger = logging.getLogger('curvefitting_Assessor')
+THRESHOLD = 0.95
+
+class CurvefittingAssessor(Assessor):
+    '''
+    CurvefittingAssessor uses learning curve fitting algorithm to predict the learning curve performance in the future.
+    It stops a pending trial X at step S if the trial's forecast result at target step is convergence and lower than the best performance in the history.
+    '''
+    def __init__(self, start_step=6, epoch_num=20):
+        if not start_step > 0:
+            logger.warning('start step should be a positive number')
+        if not epoch_num > 0:
+            logger.warning('number of epoch should be a positive number')
+        # Record the target position we predict
+        self.target_pos = epoch_num
+        # Record the best performance so far
+        self.completed_best_performance = 0.0001
+        # Start forecasting when historical data reaches start step
+        self.start_step = start_step
+        # self.start_step = max(start_step, int(epoch_num / 4))
+        logger.info('Successfully initials the curvefitting assessor')
+
+    def trial_end(self, trial_job_id, success):
+        '''
+        trial end: update the best performance of completed trial job
+        '''
+        if success:
+            self.completed_best_performance = max(self.completed_best_performance, np.max(self.trial_history))
+            logger.info('Successully update complted best performance, trial job id:', trial_job_id)
+        else:
+            logger.info('No need to update, trial job id: ', trial_job_id)
+
+    def assess_trial(self, trial_job_id, trial_history):
+        '''
+        assess whether a trial should be early stop by curve fitting
+        return AssessResult.Good or AssessResult.Bad
+        '''
+        self.trial_history = trial_history
+        curr_step = len(trial_history)
+        if curr_step < self.start_step:
+            return AssessResult.Good
+        
+        try:
+            curvemodel = CurveModel(self.target_pos)
+            predict_y = curvemodel.predict(trial_history)
+            print ("predict y ", predict_y)
+            if predict_y == -1:
+                '''wait for more information to predict precisely'''
+                return AssessResult.Good
+            elif predict_y / self.completed_best_performance > THRESHOLD:
```

Comment:
suggest to make `THRESHOLD` configurable in this assessor's spec. BTW, is 0.95 too high?

Reply:
It depends, my original purpose is to stop all the curves that can't exceed the best history performance. For example, if THRESHOLD = 0.9 and the best_performance is 0.9, then the predicted result of 0.82 will get AssessResult.Good, I think we should early stop this case.
Anyway, make THRESHOLD configurable is a good suggestion, I will update it.